File: src/main/java/com/joecollins/models/general/Candidate.java
package com.joecollins.models.general;

import java.util.Objects;

public class Candidate {

  public static final Candidate OTHERS = new Candidate("", Party.OTHERS);

  private final String name;
  private final Party party;
  private final boolean incumbent;

  private final int hash;

  public Candidate(String name, Party party) {
    this(name, party, false);
  }

  public Candidate(String name, Party party, boolean incumbent) {
    this.name = name;
    this.party = party;
    this.incumbent = incumbent;

    this.hash = Objects.hash(this.name, this.party, this.incumbent);
  }

  public String getName() {
    return name;
  }

  public Party getParty() {
    return party;
  }

  public boolean isIncumbent() {
    return incumbent;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    Candidate candidate = (Candidate) o;
    return incumbent == candidate.incumbent
        && Objects.equals(name, candidate.name)
        && Objects.equals(party, candidate.party);
  }

  @Override
  public int hashCode() {
    return hash;
  }

  @Override
  public String toString() {
    return name + " (" + party + ")" + (incumbent ? "*" : "");
  }
}

File: src/main/kotlin/com/joecollins/models/general/Candidate.kt
package com.joecollins.models.general

data class Candidate private constructor(val name: String, val party: Party, val incumbent: Boolean, override val overrideSortOrder: Int?) : CanOverrideSortOrder() {

    constructor(name: String, party: Party, incumbent: Boolean) : this(name, party, incumbent, null)

    constructor(name: String, party: Party) : this(name, party, false)

    override fun toString(): String = name + " (" + party + ")" + (if (incumbent) "*" else "")

    fun isIncumbent() = incumbent

    companion object {
        @JvmField val OTHERS = Candidate("Others", Party.OTHERS, false, -1)
    }
}
